frame/curve-amm/src/lib.rs

```diff
@@ -1011,53 +656,73 @@ pub mod pallet {
 			None
 		}
 
-		fn transfer_liquidity_into_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source: &T::AccountId,
-			destination_asset_index: usize,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[destination_asset_index],
-				source,
-				pool_account_id,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[destination_asset_index] = pool.total_balances
-				[destination_asset_index]
-				.checked_add(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
-		fn transfer_liquidity_from_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source_asset_index: usize,
-			destination: &T::AccountId,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[source_asset_index],
-				pool_account_id,
-				destination,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[source_asset_index] = pool.total_balances[source_asset_index]
-				.checked_sub(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
 		pub fn account_id(pool_id: &PoolId) -> T::AccountId {
 			PALLET_ID.into_sub_account(pool_id)
 		}
+
+		pub fn initialize_block(block_number: T::BlockNumber) {
+			with_transaction(|| {
+				let results = Pools::<T>::iter()
+					.map(|(pool_id, pool_info)| {
+						let pool_account_id = Self::account_id(&pool_id);
+						for vault_id in pool_info.assets_vault_ids {
+							let asset_id = <T::Vault>::asset_id(&vault_id)?;
+							match <T::Vault as StrategicVault>::available_funds(
+								&vault_id,
+								&pool_account_id,
+							)? {
+								FundsAvailability::Depositable(balance) => {
+									let balance = <T as Config>::LpToken::reducible_balance(
+										asset_id,
+										&pool_account_id,
+										false,
+									)
+									.min(balance);
```

Comment:
i think here you should check that ask from vault does not violates invariant to far OR (better) collect all depositables, and find maximal value of each balance you can deposit back without violation of invariant.